File: examples/rmi/HelloWorld/java/src/dds/rmi/example/helloworld/async_client.java
package dds.rmi.example.helloworld;

import java.io.*;
import DDS_RMI.HelloWorld.HelloServiceInterfaceProxy.*;

public class async_client {
    boolean done = false;

    synchronized void done() {
        done = true;
    }

    synchronized boolean isDone() {
        return done;
    }

    public static void main(String[] args) throws InterruptedException {

        String sServerName = "HelloWorldServer";
        int iProxyInstanceID = 1;
        async_client clientApp = new async_client();

        org.opensplice.DDS_RMI.Sync_Print
                .stdOutFormat("=== Getting a DDS Runtime instance\n");
        org.opensplice.DDS_RMI.CRuntime runtime = org.opensplice.DDS_RMI.CRuntime
                .getDefaultRuntime();
        if (null == runtime) {
            org.opensplice.DDS_RMI.Sync_Print
                    .stdOutFormat("*** Failed to get the DDS Runtime\n");
            System.exit(1);
        }

        org.opensplice.DDS_RMI.Sync_Print
                .stdOutFormat("=== Starting the DDS Runtime\n");
        boolean result = runtime.start(args);
        if (!result) {
            org.opensplice.DDS_RMI.Sync_Print
                    .stdOutFormat("*** Failed to start the DDS Runtime\n");
            System.exit(1);
        }

        // Interface resolution and proxy creation
        org.opensplice.DDS_RMI.Sync_Print.stdOutFormat(
                "=== Server Name=%s proxyInstanceID=%d\n", sServerName,
                iProxyInstanceID);

        DDS_RMI.HelloWorld.HelloServiceInterfaceProxy l_Proxy = null;
        try {
            l_Proxy = org.opensplice.DDS_RMI.DDS_Service.getServerProxy(
                    sServerName, iProxyInstanceID,
                    DDS_RMI.HelloWorld.HelloServiceInterfaceProxy.class);
        } catch (org.opensplice.DDS_RMI.SERVICE_NOT_FOUND e) {
            org.opensplice.DDS_RMI.Sync_Print.stdOutFormat(
                    "*** Failed to find the service '%s'\n", sServerName);
            System.exit(1);
        }

        if (null != l_Proxy) {

            try {
                org.opensplice.DDS_RMI.Sync_Print
                        .stdOutFormat("   === async_greet\n");

                async_client.MyGreetReplyHandler handler = clientApp.new MyGreetReplyHandler();
                l_Proxy.async_greet(handler);

                while (!clientApp.isDone()) {
                    Thread.sleep(2000);
                }

                org.opensplice.DDS_RMI.Sync_Print.stdOutFormat("   === stop\n");
                l_Proxy.stop();
                org.opensplice.DDS_RMI.Sync_Print
                        .stdOutFormat("   === stop done\n");
            } catch (org.opensplice.DDS_RMI.TIMEOUT e) {
                org.opensplice.DDS_RMI.Sync_Print
                        .stdOutFormat("   *** stop timed out\n");
            } catch (org.opensplice.DDS_RMI.BAD_PARAM e) {
                org.opensplice.DDS_RMI.Sync_Print
                .stdOutFormat("   *** Reply handler already active !\n");
            }

        } else {
            org.opensplice.DDS_RMI.Sync_Print
                    .stdOutFormat("*** Failed to create the proxy\n");
            System.exit(1);
        }

        org.opensplice.DDS_RMI.Sync_Print
                .stdOutFormat("=== Stopping the DDS Runtime\n");
        result = runtime.stop();
        if (!result) {
            org.opensplice.DDS_RMI.Sync_Print
                    .stdOutFormat("*** Failed to stop the DDS Runtime\n");
            System.exit(1);
        }
    }

    /**
     * Reply Handler of the 'greet' operation
     *
     */
    class MyGreetReplyHandler extends greet_Reply_Handler {
        public void greet_Reply(String ret) {
            System.out.println("async_greet returns: " + ret);
            done();
        }
    };
}

File: examples/rmi/HelloWorld/cpp/src/async_client.cpp
// For Windows build
#if defined(WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(__CYGWIN__)
#include "WindowsBuild.h"
#else
#define DDS_SERVICE_API
#endif

#include "ddsrmi.h"
#include "HelloWorld_InterfaceProxy.h"

#include "example_main.h"

using namespace org::opensplice::DDS_RMI;

/**
 * @brief A DDS RMI Client Example using asynchronous invocations
 */

bool done;

/**
 * Reply Handler of the 'greet' operation
 *
 */
class MyGreetReplyHandler : public virtual DDS_RMI::HelloWorld::HelloService_greet_Reply_Handler
{
  void
  greet_Reply(DDS::String ret)
  {
    std::cout << "async_greet returns: " << ret << std::endl;
    done = true;
  }
};

int
OSPL_MAIN(int argc, const char * argv[])
{

  std::string sServerName("HelloWorldServer");
  int iProxyInstanceID = 1;
  done = false;

  // Getting a DDS Runtime instance
  std::cout << "=== getDefaultRuntime" << std::endl;
  CRuntime_ref runtime = CRuntime::getDefaultRuntime();
  if (runtime.get() == NULL)
    {
      std::cout << "*** Client main: Failed to get the default Runtime "
          << std::endl;
    }

  // Starting the DDS runtime
  std::cout << "=== Runtime start" << std::endl;
  bool result = runtime->start(argc, argv);
  if (result != true)
    {
      std::cout << "*** Client main: Failed to start the Runtime " << std::endl;
    }

  std::cout << "=== Getting the server proxy ..." << std::endl;
  shared_ptr< ::DDS_RMI::HelloWorld::HelloServiceInterfaceProxy> cproxy;
  try
    {
      bool ret = DDS_Service::getServerProxy<
          ::DDS_RMI::HelloWorld::HelloServiceInterfaceProxy>(sServerName,
          iProxyInstanceID, cproxy);
      if (!ret)
        {
          std::cout << " *** Failed to get the server proxy !" << std::endl;
        }
      else
        {
          std::cout << "=== Succeeded in getting the server proxy named "
              << sServerName << " : " << std::endl;
        }
    }
  catch (SERVICE_NOT_FOUND&)
    {
      std::cout << " *** Unable to find the service !" << std::endl;
    }

  try
    {
      shared_ptr<MyGreetReplyHandler> handler(new MyGreetReplyHandler());
      std::cout << "=== async_greet" << std::endl;
      cproxy->async_greet(handler.get());
    }
  catch (BAD_PARAM)
    {
      std::cout << "*** Operation TIMEOUT" << std::endl;
    }

  os_time delay_2s = { 2, 0 };
  while(!done){
      os_nanoSleep(delay_2s);
  }

  try
    {
      std::cout << "=== stop" << std::endl;
      cproxy->stop();
      std::cout << "=== stop done" << std::endl;
    }
  catch (TIMEOUT&)
    {
      std::cout << "*** Operation TIMEOUT" << std::endl;
    }

  // Stopping the Runtime
  std::cout << "=== Runtime stop" << std::endl;
  runtime->stop();
  return 0;
}
